File: apps/mobile/src/lib/colonies.ts
/**
 * Colony mode API client — ADR-010.
 *
 * Population-level tracking for communal/colony keepers (roaches, millipedes,
 * isopods/springtails-as-"other"). A colony is a first-class collection entry
 * with per-life-stage headcounts, tracked as ONE entry (counts 1 toward the
 * free cap regardless of headcount).
 *
 * Modeled on the FeederColony pattern (JSONB stage buckets + an events stream
 * that adjusts the buckets). The taxon vocabulary is the SAME shared invert
 * registry (INVERT_TAXA / INVERT_TAXON_ORDER) — colonies do not invent a
 * parallel taxon list.
 *
 * apiClient baseURL already includes /api/v1 — paths start at /colonies.
 * Collection endpoints use a trailing slash; by-id/action endpoints do not.
 */
import { apiClient } from '../services/api';
import type { InvertTaxon, Source, Visibility } from './inverts';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Per-life-stage buckets. Casual keepers can dump everything in "mixed". */
export type StageCounts = Record<string, number>;

export type ColonyEventType =
  | 'birth'
  | 'death'
  | 'added'
  | 'removed'
  | 'cannibalism'
  | 'aggression'
  | 'molt_found'
  | 'split'
  | 'merge'
  | 'observation'
  | 'count_correction';

export interface ColonyListItem {
  id: string;
  taxon: InvertTaxon;
  name: string;
  photo_url: string | null;
  total_count: number;
  count_is_estimated: boolean;
  stage_counts: StageCounts;
  is_active: boolean;
  species_display_name: string | null;
  species_scientific_name: string | null;
  species_missing: boolean;
  /** Last ACCEPTED feeding. Lets the collection card read "Fed 4d ago" like
   *  every other card. */
  last_feeding_date?: string | null;
  days_since_last_feeding?: number | null;
}

export interface Colony {
  id: string;
  user_id: string;
  taxon: InvertTaxon;
  species_id: string | null;
  enclosure_id: string | null;
  name: string;
  date_acquired: string | null;
  founded_date: string | null;
  source: Source | null;
  stage_counts: StageCounts;
  count_is_estimated: boolean;
  /** terrestrial | arboreal | fossorial. Same vocabulary as inverts. */
  enclosure_type: string | null;
  /** Free text, e.g. "12x12x12 inches". Floor space per animal is the main
   *  driver of communal success, so this matters more here than on a solitary
   *  animal — it was missing entirely until 2026-07-29. */
  enclosure_size: string | null;
  substrate_type: string | null;
  substrate_depth: string | null;
  last_substrate_change: string | null;
  target_temp_min: string | null;
  target_temp_max: string | null;
  target_humidity_min: string | null;
  target_humidity_max: string | null;
  water_dish: boolean;
  notes: string | null;
  photo_url: string | null;
  visibility: Visibility | null;
  is_active: boolean;
  created_at: string;
  updated_at: string | null;
  total_count: number;
  species_display_name: string | null;
  species_scientific_name: string | null;
  species_missing: boolean;
}

export interface ColonyCreate {
  name: string;
  taxon: InvertTaxon;
  species_id?: string | null;
  enclosure_id?: string | null;
  date_acquired?: string | null;
  founded_date?: string | null;
  source?: Source | null;
  stage_counts?: StageCounts | null;
  count_is_estimated?: boolean;
  enclosure_type?: string | null;
  enclosure_size?: string | null;
  substrate_type?: string | null;
  substrate_depth?: string | null;
  last_substrate_change?: string | null;
  target_temp_min?: string | null;
  target_temp_max?: string | null;
  target_humidity_min?: string | null;
  target_humidity_max?: string | null;
  water_dish?: boolean;
  notes?: string | null;
  photo_url?: string | null;
  visibility?: Visibility | null;
}

/** Partial update — everything optional (name/taxon not required on PUT). */
export type ColonyUpdate = Partial<Omit<ColonyCreate, 'taxon'>> & {
  taxon?: InvertTaxon;
  is_active?: boolean;
};

export interface ColonyEvent {
  id: string;
  colony_id: string;
  user_id: string;
  event_type: ColonyEventType;
  stage: string | null;
  count_delta: number | null;
  occurred_at: string; // YYYY-MM-DD
  severity: string | null;
  notes: string | null;
  created_at: string;
}

export interface ColonyEventCreate {
  event_type: ColonyEventType;
  stage?: string | null;
  count_delta?: number | null;
  occurred_at?: string | null;
  severity?: string | null;
  notes?: string | null;
}

export type ColonyEventUpdate = Partial<ColonyEventCreate>;

// ---------------------------------------------------------------------------
// Colony CRUD
// ---------------------------------------------------------------------------

export async function listColonies(includeInactive = false): Promise<ColonyListItem[]> {
  const { data } = await apiClient.get<ColonyListItem[]>('/colonies/', {
    params: { include_inactive: includeInactive },
  });
  return data;
}

export async function getColony(id: string): Promise<Colony> {
  const { data } = await apiClient.get<Colony>(`/colonies/${id}`);
  return data;
}

export async function createColony(payload: ColonyCreate): Promise<Colony> {
  const { data } = await apiClient.post<Colony>('/colonies/', payload);
  return data;
}

export async function updateColony(id: string, payload: ColonyUpdate): Promise<Colony> {
  const { data } = await apiClient.put<Colony>(`/colonies/${id}`, payload);
  return data;
}

export async function deleteColony(id: string): Promise<void> {
  await apiClient.delete(`/colonies/${id}`);
}

// ---------------------------------------------------------------------------
// Colony events (population lifecycle — adjusts stage_counts on write)
// ---------------------------------------------------------------------------

export async function listColonyEvents(id: string): Promise<ColonyEvent[]> {
  const { data } = await apiClient.get<ColonyEvent[]>(`/colonies/${id}/events`);
  return data;
}

export async function createColonyEvent(id: string, payload: ColonyEventCreate): Promise<ColonyEvent> {
  const { data } = await apiClient.post<ColonyEvent>(`/colonies/${id}/events`, payload);
  return data;
}

export async function updateColonyEvent(eventId: string, payload: ColonyEventUpdate): Promise<ColonyEvent> {
  const { data } = await apiClient.put<ColonyEvent>(`/colonies/events/${eventId}`, payload);
  return data;
}

export async function deleteColonyEvent(eventId: string): Promise<void> {
  await apiClient.delete(`/colonies/events/${eventId}`);
}

// ---------------------------------------------------------------------------
// Display helpers
// ---------------------------------------------------------------------------

/** Sum a stage_counts object into a total headcount. */
export function sumStageCounts(counts: StageCounts | null | undefined): number {
  if (!counts) return 0;
  return Object.values(counts).reduce((acc, n) => acc + (Number.isFinite(n) ? n : 0), 0);
}

/** "≈N" when estimated, "N" otherwise. */
export function formatColonyCount(total: number, estimated: boolean): string {
  const n = total.toLocaleString();
  return estimated ? `≈${n}` : n;
}

/** Human-readable label for an event type. */
export const COLONY_EVENT_LABELS: Record<ColonyEventType, string> = {
  birth: 'Birth',
  death: 'Death',
  added: 'Added',
  removed: 'Removed',
  cannibalism: 'Cannibalism',
  aggression: 'Aggression',
  molt_found: 'Molt found',
  split: 'Split',
  merge: 'Merge',
  observation: 'Observation',
  count_correction: 'Count correction',
};

/** Emoji glyph for an event type — used in the timeline. */
export const COLONY_EVENT_ICONS: Record<ColonyEventType, string> = {
  birth: '🐣',
  death: '💀',
  added: '➕',
  removed: '➖',
  cannibalism: '🩸',
  aggression: '⚔️',
  molt_found: '🐚',
  split: '🔀',
  merge: '🔗',
  observation: '📝',
  count_correction: '✏️',
};

/** Event types that carry a severity rating (aggression/cannibalism). */
export function eventHasSeverity(t: ColonyEventType): boolean {
  return t === 'aggression' || t === 'cannibalism';
}

// ---------------------------------------------------------------------------
// Photos + feedings (cph_20260729_colony_logs)
//
// A communal tarantula is a display animal housed as a group, so it gets the
// same gallery and feeding log as any other animal. Feeder colonies can use
// these too — nothing is withheld — but the UI doesn't push them at it.
// ---------------------------------------------------------------------------

export interface ColonyPhoto {
  id: string;
  url: string;
  thumbnail_url: string | null;
  caption: string | null;
  taken_at: string | null;
  created_at: string;
}

export interface ColonyFeedingLog {
  id: string;
  colony_id: string | null;
  fed_at: string;
  food_type: string | null;
  food_size: string | null;
  /** How many prey items went in. THE number for a group feeding — you don't
   *  feed a communal "a cricket", you drop six in for eleven spiders, and the
   *  ratio is what tells you whether the colony is being fed enough. */
  quantity: number | null;
  /** For a colony this means "did the GROUP take it", not any one animal —
   *  a communal is fed as a unit and you can't see which spider ate what. */
  accepted: boolean;
  notes: string | null;
}

/**
 * A molt found in the colony.
 *
 * Unattributed by definition — you can't tell which of eleven spiders shed it,
 * which is exactly why `is_unidentified` is forced true server-side. The
 * measurement fields exist because the table is shared with individual animals;
 * for a colony they're expected to stay null, since you can't weigh a molt's
 * owner when you don't know who that is.
 *
 * Worth more here than for a solitary animal, not less: for a communal a shed
 * skin is often the only observation that surfaces on its own, and sexing a
 * communal means sexing molts.
 */
export interface ColonyMoltLog {
  id: string;
  colony_id: string | null;
  molted_at: string;
  notes: string | null;
  image_url: string | null;
  is_unidentified: boolean;
}

export async function listColonyMolts(id: string): Promise<ColonyMoltLog[]> {
  const { data } = await apiClient.get<ColonyMoltLog[]>(`/colonies/${id}/molts`);
  return data;
}

export async function createColonyMolt(
  id: string,
  payload: { molted_at: string; notes?: string | null; image_url?: string | null },
): Promise<ColonyMoltLog> {
  const { data } = await apiClient.post<ColonyMoltLog>(`/colonies/${id}/molts`, payload);
  return data;
}

/** Shared with individual animals — molts resolve ownership through whichever
 *  parent they carry, so no colony-specific delete route was needed. */
export async function deleteColonyMolt(moltId: string): Promise<void> {
  await apiClient.delete(`/molts/${moltId}`);
}

export async function listColonyPhotos(id: string): Promise<ColonyPhoto[]> {
  const { data } = await apiClient.get<ColonyPhoto[]>(`/colonies/${id}/photos`);
  return data;
}

export async function listColonyFeedings(id: string): Promise<ColonyFeedingLog[]> {
  const { data } = await apiClient.get<ColonyFeedingLog[]>(`/colonies/${id}/feedings`);
  return data;
}

export async function createColonyFeeding(
  id: string,
  payload: {
    fed_at: string;
    food_type?: string | null;
    food_size?: string | null;
    quantity?: number | null;
    accepted?: boolean;
    notes?: string | null;
  },
): Promise<ColonyFeedingLog> {
  const { data } = await apiClient.post<ColonyFeedingLog>(`/colonies/${id}/feedings`, payload);
  return data;
}

/** Promote a photo to the colony's hero. Shares the generic photo route —
 *  `set-main` resolves ownership through the photo's parent, so it needed no
 *  colony-specific handling. */
export async function setColonyMainPhoto(photoId: string): Promise<void> {
  await apiClient.patch(`/photos/${photoId}/set-main`);
}

export async function deleteColonyPhoto(photoId: string): Promise<void> {
  await apiClient.delete(`/photos/${photoId}`);
}

export async function uploadColonyPhoto(id: string, form: FormData): Promise<ColonyPhoto> {
  const { data } = await apiClient.post<ColonyPhoto>(`/colonies/${id}/photos`, form, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return data;
}

File: apps/api/app/routers/colonies.py
"""
Colony + colony-event routes (ADR-010).

Population-level tracking for communal/colony keepers. Owner-scoped throughout.
A colony counts as 1 toward the free-tier animal cap (enforced on create via
the shared `enforce_collection_limit`, which counts inverts + colonies).

Events with a `count_delta` mutate the colony's `stage_counts` bucket on write
(same pattern as FeederCareLog -> count). JSONB is reassigned (not mutated in
place) so SQLAlchemy flushes the change.
"""
from typing import List, Optional
from uuid import UUID
from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.user import User
from app.models.colony import Colony, ColonyEvent
from app.models.invert_species import InvertSpecies
from app.models.enclosure import Enclosure
from app.schemas.colony import (
    ColonyCreate,
    ColonyUpdate,
    ColonyResponse,
    ColonyListItem,
    ColonyEventCreate,
    ColonyEventUpdate,
    ColonyEventResponse,
)
from app.utils.dependencies import get_current_user
from app.utils.limits import enforce_collection_limit

router = APIRouter()


# ---------- helpers ----------

def _species_names(sp: Optional[InvertSpecies]):
    """(display_name, scientific_name). Display prefers a common name."""
    if sp is None:
        return None, None
    display = sp.common_names[0] if sp.common_names else sp.scientific_name
    return display, sp.scientific_name


def _total_count(colony: Colony) -> int:
    """Sum the buckets. Always an int (0 for an empty colony) so the clients
    can type total_count as a number without a null branch."""
    if not colony.stage_counts:
        return 0
    try:
        return sum(int(v) for v in colony.stage_counts.values() if isinstance(v, int))
    except Exception:
        return 0


def _build_response(colony: Colony, db: Session) -> dict:
    sp: Optional[InvertSpecies] = None
    species_missing = False
    if colony.species_id:
        sp = db.query(InvertSpecies).filter(InvertSpecies.id == colony.species_id).first()
        if sp is None:
            species_missing = True

    display, scientific = _species_names(sp)
    data = {c.name: getattr(colony, c.name) for c in colony.__table__.columns}
    data["total_count"] = _total_count(colony)
    data["species_display_name"] = display
    data["species_scientific_name"] = scientific
    data["species_missing"] = species_missing
    return data


def _verify_enclosure(db: Session, enclosure_id: Optional[UUID], user: User) -> None:
    if enclosure_id is None:
        return
    enc = (
        db.query(Enclosure)
        .filter(Enclosure.id == enclosure_id, Enclosure.user_id == user.id)
        .first()
    )
    if enc is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Enclosure not found")


def _verify_species(db: Session, species_id: Optional[UUID]) -> None:
    if species_id is None:
        return
    sp = db.query(InvertSpecies).filter(InvertSpecies.id == species_id).first()
    if sp is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Species not found")


def _get_owned(db: Session, colony_id: UUID, user: User) -> Colony:
    colony = (
        db.query(Colony)
        .filter(Colony.id == colony_id, Colony.user_id == user.id)
        .first()
    )
    if colony is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Colony not found")
    return colony


def _apply_delta(colony: Colony, stage: Optional[str], delta: Optional[int]) -> None:
    """Adjust a bucket by delta (clamped at 0). Reassigns the JSONB dict so
    SQLAlchemy detects the change. Missing stage defaults to 'mixed'."""
    if delta is None:
        return
    bucket = (stage or "mixed").strip() or "mixed"
    counts = dict(colony.stage_counts or {})
    counts[bucket] = max(0, int(counts.get(bucket, 0)) + int(delta))
    colony.stage_counts = counts


# ---------- colony CRUD ----------

@router.get("/", response_model=List[ColonyListItem])
async def list_colonies(
    include_inactive: bool = Query(False),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    q = db.query(Colony).filter(
        Colony.user_id == current_user.id,
        Colony.transferred_out_at.is_(None),
    )
    if not include_inactive:
        q = q.filter(Colony.is_active.is_(True))
    colonies = q.order_by(Colony.created_at.desc()).all()
    return [ColonyListItem(**_build_response(c, db)) for c in colonies]


@router.post("/", response_model=ColonyResponse, status_code=status.HTTP_201_CREATED)
async def create_colony(
    payload: ColonyCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    # A colony counts as 1 animal toward the free-tier cap.
    enforce_collection_limit(db, current_user)
    _verify_enclosure(db, payload.enclosure_id, current_user)
    _verify_species(db, payload.species_id)

    colony = Colony(user_id=current_user.id, **payload.model_dump())
    db.add(colony)

    # Bump the species "times_kept" counter (parity with invert create).
    if colony.species_id:
        sp = db.query(InvertSpecies).filter(InvertSpecies.id == colony.species_id).first()
        if sp is not None:
            sp.times_kept = (sp.times_kept or 0) + 1

    db.commit()
    db.refresh(colony)
    return ColonyResponse(**_build_response(colony, db))


@router.get("/{colony_id}", response_model=ColonyResponse)
async def get_colony(
    colony_id: UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    colony = _get_owned(db, colony_id, current_user)
    return ColonyResponse(**_build_response(colony, db))


@router.put("/{colony_id}", response_model=ColonyResponse)
async def update_colony(
    colony_id: UUID,
    payload: ColonyUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    colony = _get_owned(db, colony_id, current_user)
    data = payload.model_dump(exclude_unset=True)

    if "enclosure_id" in data:
        _verify_enclosure(db, data["enclosure_id"], current_user)
    if data.get("species_id"):
        _verify_species(db, data["species_id"])

    for k, v in data.items():
        setattr(colony, k, v)

    db.commit()
    db.refresh(colony)
    return ColonyResponse(**_build_response(colony, db))


@router.delete("/{colony_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_colony(
    colony_id: UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    colony = _get_owned(db, colony_id, current_user)
    db.delete(colony)
    db.commit()
    return None


# ---------- events ----------

@router.get("/{colony_id}/events", response_model=List[ColonyEventResponse])
async def list_events(
    colony_id: UUID,
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    _get_owned(db, colony_id, current_user)  # ownership check
    return (
        db.query(ColonyEvent)
        .filter(ColonyEvent.colony_id == colony_id)
        .order_by(ColonyEvent.occurred_at.desc(), ColonyEvent.created_at.desc())
        .offset(offset)
        .limit(limit)
        .all()
    )


@router.post(
    "/{colony_id}/events",
    response_model=ColonyEventResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_event(
    colony_id: UUID,
    payload: ColonyEventCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    colony = _get_owned(db, colony_id, current_user)

    log = ColonyEvent(
        colony_id=colony.id,
        user_id=current_user.id,
        **payload.model_dump(exclude_unset=True),
    )
    db.add(log)

    # Apply the population delta to the bucket.
    _apply_delta(colony, log.stage, log.count_delta)

    db.commit()
    db.refresh(log)
    return log


@router.put("/events/{event_id}", response_model=ColonyEventResponse)
async def update_event(
    event_id: UUID,
    payload: ColonyEventUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    log = (
        db.query(ColonyEvent)
        .filter(ColonyEvent.id == event_id, ColonyEvent.user_id == current_user.id)
        .first()
    )
    if log is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Event not found")

    data = payload.model_dump(exclude_unset=True)
    for k, v in data.items():
        setattr(log, k, v)

    db.commit()
    db.refresh(log)
    return log


@router.delete("/events/{event_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_event(
    event_id: UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    log = (
        db.query(ColonyEvent)
        .filter(ColonyEvent.id == event_id, ColonyEvent.user_id == current_user.id)
        .first()
    )
    if log is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Event not found")

    db.delete(log)
    db.commit()
    return None
